### DISCUSSION TITLE
How to instantiate plugins with LightningCLI?

### DISCUSSION THREAD

**Discussion by awaelchli** [Lightning-AI/pytorch-lightning#8561] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
I'm looking at the following example: 

```py
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule
from pytorch_lightning.utilities.cli import LightningCLI


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=2)


def run():
    LightningCLI(model_class=BoringModel)


if __name__ == "__main__":
    run()
```

with config file: 

```
trainer:
    max_steps: 1
    limit_val_batches: 1
    accelerator: ddp_cpu
    plugins:
      - class_path: "pytorch_lightning.plugins.training_type.ddp.DDPPlugin"
        init_args:
          find_unused_parameters: false

```

In Lightning 1.4 I am getting the following error.

```python
bug_report_model.py: error: Parser key "trainer.plugins": Value "[{'class_path': 'pytorch_lightning.plugins.training_type.ddp.DDPPlugin', 'init_args': {'find_unused_parameters': True}}]" does not validate against any of the types in typing.Union[typing.List[typing.Union[pytorch_lightning.plugins.base_plugin.Plugin, pytorch_lightning.plugins.environments.cluster_environment.ClusterEnvironment, str]], pytorch_lightning.plugins.base_plugin.Plugin, pytorch_lightning.plugins.environments.cluster_environment.ClusterEnvironment, str, NoneType]:
  - Value "{'class_path': 'pytorch_lightning.plugins.training_type.ddp.DDPPlugin', 'init_args': {'find_unused_parameters': True}}" does not validate against any of the types in typing.Union[pytorch_lightning.plugins.base_plugin.Plugin, pytorch_lightning.plugins.environments.cluster_environment.ClusterEnvironment, str]:
    - Problem with given class_path "pytorch_lightning.plugins.training_type.ddp.DDPPlugin":
      - 'Configuration check failed :: No action for key "find_unused_parameters" to check its value.'
    - "pytorch_lightning.plugins.training_type.ddp.DDPPlugin" is not a subclass of ClusterEnvironment
    - Expected a <class 'str'> but got "{'class_path': 'pytorch_lightning.plugins.training_type.ddp.DDPPlugin', 'init_args': {'find_unused_parameters': True}}"
  - Type <class 'pytorch_lightning.plugins.base_plugin.Plugin'> expects an str or a Dict with a class_path entry but got "[{'class_path': 'pytorch_lightning.plugins.training_type.ddp.DDPPlugin', 'init_args': {'find_unused_parameters': True}}]"
  - Type <class 'pytorch_lightning.plugins.environments.cluster_environment.ClusterEnvironment'> expects an str or a Dict with a class_path entry but got "[{'class_path': 'pytorch_lightning.plugins.training_type.ddp.DDPPlugin', 'init_args': {'find_unused_parameters': True}}]"
  - Expected a <class 'str'> but got "[{'class_path': 'pytorch_lightning.plugins.training_type.ddp.DDPPlugin', 'init_args': {'find_unused_parameters': True}}]"
  - Expected a <class 'NoneType'> but got "[{'class_path': 'pytorch_lightning.plugins.training_type.ddp.DDPPlugin', 'init_args': {'find_unused_parameters': True}}]"
```

The relevant error is:

```python
    - Problem with given class_path "pytorch_lightning.plugins.training_type.ddp.DDPPlugin":
      - 'Configuration check failed :: No action for key "find_unused_parameters" to check its value.'
```

The [DDPPlugin signature has kwargs defined](https://github.com/PyTorchLightning/pytorch-lightning/blob/75e18a52988f760a8b32a3c4fc6f0f3217868e57/pytorch_lightning/plugins/training_type/ddp.py#L74) and I am not sure if jsonargparse is supposed to be able to handle them, I couldn't find any explanation in their [docs](https://jsonargparse.readthedocs.io/en/stable/#class-type-and-sub-classes). 

I would like to pass in `find_unused_parameters=False` as part of these kwargs. 

Current workaround: 

```
trainer:
    ...
    plugins: ddp_find_unused_parameters_false 

```

cc @mauvilsa @carmocca who might be pleased to hear that I started to really like the CLI <3  

EDIT
- A similar issue exists here #8262, but it is not the same because what OP posted works with the latest jsonargparse version and is not my exact use case.

**Comment by mauvilsa** [comment]
There are many possible ways in which people could use `**kwargs` in the `__init__` of a class. Having `jsonargparse` try to interpret the source code to see how `**kwargs` is used would be highly complex. Thus it is a design decision that `**kwargs` is used in a single way, and that is for class inheritance. The documentation does mention it "since the init has the `**kwargs` argument, the keyword arguments from MyBaseClass are also added". This might be a bit too brief and could be good to extend it. Also one objective of `jsonargparse` has been to validate everything so that input mistakes make the code fail as early as possible with useful messages. Having `**kwargs` mean to accept anything goes against this.

Is there any class or function that defines the accepted parameters and type hints for the `DDPPlugin` init kwargs?

**Reply by awaelchli** [reply]
Hi @mauvilsa 
That certainly makes sense and I agree it would contradict the input validation that jsonargparse provides. I also think that it's always dangerous to have kwargs in the signature since typos in named inputs could be silently land in the kwargs collection. 

> Having jsonargparse try to interpret the source code to see how **kwargs is used would be highly complex

No, that's not possible. This is out of the question. I'm not asking this for sure 😅 

> Is there any class or function that defines the accepted parameters and type hints for the DDPPlugin init kwargs?

Yes, in the case of the plugin the kwargs are saved and then later forwarded directly to this class here: 
https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html 

https://github.com/PyTorchLightning/pytorch-lightning/blob/eaa16c7480ef6728d69113d6305014d39c2efece/pytorch_lightning/plugins/training_type/ddp.py#L101

https://github.com/PyTorchLightning/pytorch-lightning/blob/eaa16c7480ef6728d69113d6305014d39c2efece/pytorch_lightning/plugins/training_type/ddp.py#L306

But I don't think it is possible for us to reflect such a validation in a type hint. So I see two possible solutions:

1) Let our DDPPlugin explicitly list the kwargs it can accept with type hints. Pro: works with LightningCLI, con: Not acnostic to pytorch's future updates to the DDP wrapper. 
2) do nothing. simply don't support plugin creation via cli and let users have pass in strings only

**Comment by quancs** [comment]
Hi @awaelchli @mauvilsa @carmocca, what about implementing `jsonargparse` to support such kind of input:
```
fit:
  trainer:
    plugins:
      - class_path: pytorch_lightning.plugins.DDPPlugin
        init_args:
          kwargs:
            find_unused_parameters: false
```
That is to regard `kwargs` as a special `Dict` that doesn't need to check the types.
It's a good compromise for the type validation and the practical needs for the `**kwargs`.

**Reply by mauvilsa** [reply]
Having a special keyword to skip the check could be a good compromise given the limitation of the specific way in which `**kwargs` is handled. Instead of calling it `kwargs` better to be like `__kwargs__` to be more explicit that it is a special keyword. Adding new keywords is something that in general I will try to avoid and will not rush to do. Though this might be a case in which there is a good justification for it.

**Reply by quancs** [reply]
I didn't explain my idea clearly. I was saying we could regard the parameters with `**` prefix as a dict that doesn't need type checking. For example, in the following code, the `**` describes `any_name`, so `any_name` is a dict that does not need type checking. 

```python
def test_kwargs(**any_name):
    print(type(any_name))
    print(any_name)

test_kwargs(a='123',b=2)
```

And as we could see from its output below, the `any_name` is indeed a dict.
```
<class 'dict'>
{'a': '123', 'b': 2}
```

To construct such sepcial `any_name`, we could specify its keys and values like the config followed without intervening the type checking of other paramters:
```
a_class_path:
  any_name: 
    a: '123'
    b: 2
  other_parameters_that_need_type_checking
```

**Reply by quancs** [reply]
@mauvilsa I think a handle for `**` described parameters is necessary, as we could never force the users to not have that kind of parameters in python. **DDPPlugin** and [**PyTorchProfiler**](https://github.com/PyTorchLightning/pytorch-lightning/issues/10859#issuecomment-1001949302) are two examples for the feature requested.

**Reply by quancs** [reply]
One more comment, the `ddp_find_unused_parameters_false` is really not an elegent approach, and it is not general too.

**Reply by carmocca** [reply]
@quancs Would you be interested in gathering your thoughts and creating a proper proposal in https://github.com/omni-us/jsonargparse/issues? It's best to move this discussion there so that this has visibility for other jsonargparse users.

> One more comment, the ddp_find_unused_parameters_false is really not an elegent approach, and it is not general too.

True. This is just the current workaround.

**Reply by mauvilsa** [reply]
@quancs using `any_name` has the downside that it forces people to know what is the particular name that was used internally in the code. But true that it guarantees that there is no clash with the names of other parameters. Making it a fixed name does risk a clash (though highly unlikely for `__*__`), but has advantages. If people see it being used, they can easily find in the documentation what it does. The actual name to use I am not yet sure. `__kwargs__` is one option, but better if it is more scope specific. This would only be for parameters that can't be resolved via class inheritance, so another option could be `__unresolved__`.

**Reply by quancs** [reply]
Hi, this disscussion has been moved to https://github.com/omni-us/jsonargparse/issues/114

**Comment by mauvilsa** [comment, marked as answer]
The issues related to plugins instantiation have been resolved with #13283.

**Reply by awaelchli** [reply]
Wow! Nice work <3
